# 单样本率的检验

215名病人中39名被观测到患有哮喘，然后有人对“随机病人”患有哮喘的概率是0.15这个假设做检验。  

In [1]:
prop.test(39,215,0.15)


	1-sample proportions test with continuity correction

data:  39 out of 215, null probability 0.15
X-squared = 1.425, df = 1, p-value = 0.2326
alternative hypothesis: true p is not equal to 0.15
95 percent confidence interval:
 0.1335937 0.2408799
sample estimates:
        p 
0.1813953 


三个参数分别是阳性观测数，总数，以及参考概率值。最后一个参数如果没有设定的话，那么默认是0.5。  

从当前的结果中得到的结论是无差别（P>0.05），而且在95%置信区间里可以看出范围包含了0.15，所以也可认为无差别。  

## 二项分布检验

能得到精确的检验概率，所以一般prop.test()应用更广泛。  

In [2]:
binom.test(39,215,0.15)


	Exact binomial test

data:  39 and 215
number of successes = 39, number of trials = 215, p-value = 0.2135
alternative hypothesis: true probability of success is not equal to 0.15
95 percent confidence interval:
 0.1322842 0.2395223
sample estimates:
probability of success 
             0.1813953 


在检验水准为0.05水平下的置信区间其实是从0.025水平下的双侧检验中得到的。  
就是说两侧分别是0.025和0.975的水平。  

# 两独立样本比例检验

函数prop.test()也能够用于比较两个或多个比例。  
这种情况下它的参数就应该是两个向量，前一个向量里存放各组阳性观测数，后一个向量里是每组总数。

继续上述案例，可以理解成两位医生手术成功率比较。  

In [3]:
success<-c(9,4)

In [4]:
total<-c(12,13)

In [5]:
prop.test(success,total)


	2-sample test for equality of proportions with continuity correction

data:  success out of total
X-squared = 3.2793, df = 1, p-value = 0.07016
alternative hypothesis: two.sided
95 percent confidence interval:
 0.01151032 0.87310506
sample estimates:
   prop 1    prop 2 
0.7500000 0.3076923 


R会将数据自动识别成两样本率的差异检验，求的是阳性结果的概率差异。

然后我们得到的P值是0.07016，这个值与0.05很近，得到的是无差异的阴性结果。

但是95%的置信区间为[0.011531032,0.87310506]这个范围没有包含0，  
这个置信区间是比例之差的置信区间，它的结论是不可以认为两个医生的手术成功率是一样的阳性结果，  
二者的差异是由置信区间和假设检验使用的是不同的近似方法导致的。  

## Fisher 检验

Fisher确切概率法的结果是在这个四格表里出现当前数据情况或者更有利于阳性结果的概率之和，  
它会计算每一种情况的概率，然后再累加起来。  

In [6]:
a<-matrix(c(9,4,3,9),nrow=2)

In [7]:
a

[,1] [,2]
[1,] 9    3   
[2,] 4    9

In [8]:
fisher.test(a)


	Fisher's Exact Test for Count Data

data:  a
p-value = 0.04718
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
  0.9006803 57.2549701
sample estimates:
odds ratio 
  6.180528 


置信区间是比值比的结果，也就是计算(p1/(1-p1))/(p2/(1-p2))的区间，  
是一个衡量Fisher检验中相关程度的指标，得到的结果可以跟1比较。  
不过这里的结果同样和假设检验的结果相矛盾，原因同上。  

## 卡方检验

chisq.test()中的标准χ2检验需要矩阵类型的数据源。  
而作为一个2*2表格来说，这个检验与prop.test()的结果是完全一致的。  

In [9]:
chisq.test(a)


	Pearson's Chi-squared test with Yates' continuity correction

data:  a
X-squared = 3.2793, df = 1, p-value = 0.07016


# 趋势检验-多组率

有的时候你想要比较多于两个部分，而这些部分又是有序的分类，所以你想找到一个随着分组序号递增或者递减的趋势。

一组数据记录了一组女性是否使用剖腹产生育孩子，以及对应产妇鞋子码数的数据，数据在R语言ISwR数据包里。

# r × c表格

为了分析行列都多于两个分类的表格数据，可以使用函数chisq.test()和 fisher.test()，  
但是后者在每一格数字比较大而且超出两行或两列时的计算量非常大。

# 参考

[R语言系列第四期：③R语言表格数据率的比较](https://www.sohu.com/a/332349854_120098972)